token-lending/program/src/state.rs

```diff
@@ -637,7 +701,130 @@ mod test {
+                cumulative_borrow_rate_wads: Decimal::zero(),
+                ..Obligation::default()
+            }
+            .accrue_interest(Decimal::one()),
+            Err(LendingError::MathOverflow.into())
+        );
+
+        assert_eq!(
+            Obligation {
+                cumulative_borrow_rate_wads: Decimal::from(2u64),
+                ..Obligation::default()
+            }
+            .accrue_interest(Decimal::one()),
+            Err(LendingError::NegativeInterestRate.into())
+        );
+
+        assert_eq!(
+            Obligation {
+                cumulative_borrow_rate_wads: Decimal::one(),
+                borrowed_liquidity_wads: Decimal::from(u64::MAX),
+                ..Obligation::default()
+            }
+            .accrue_interest(Decimal::from(10 * MAX_COMPOUNDED_INTEREST)),
+            Err(LendingError::MathOverflow.into())
+        );
+    }
+
+    // Creates rates (r1, r2) where 0 < r1 <= r2 <= 100*r1
+    fn cumulative_rates() -> impl Strategy<Value = (u128, u128)> {
+        prop::num::u128::ANY.prop_flat_map(|rate| {
+            let current_rate = rate.max(1);
+            let max_new_rate = current_rate.saturating_mul(MAX_COMPOUNDED_INTEREST as u128);
+            (Just(current_rate), current_rate..=max_new_rate)
+        })
+    }
+
     proptest! {
+        #[test]
+        fn current_utilization_rate(
+            total_liquidity in 0..=MAX_LIQUIDITY,
+            borrowed_percent in 0..=WAD,
+        ) {
+            let borrowed_liquidity_wads = Decimal::from(total_liquidity).try_mul(Rate::from_scaled_val(borrowed_percent))?;
+            let available_liquidity = total_liquidity - borrowed_liquidity_wads.try_round_u64()?;
+            let state = ReserveState {
+                borrowed_liquidity_wads,
+                available_liquidity,
+                ..ReserveState::default()
+            };
+
+            let current_rate = state.current_utilization_rate()?;
+            assert!(current_rate <= Rate::one());
+        }
+
+        #[test]
+        fn collateral_exchange_rate(
+            total_liquidity in 0..=MAX_LIQUIDITY,
+            borrowed_percent in 0..=WAD,
+            collateral_multiplier in 0..=(5*WAD),
+        ) {
+            let borrowed_liquidity_wads = Decimal::from(total_liquidity).try_mul(Rate::from_scaled_val(borrowed_percent))?;
+            let available_liquidity = total_liquidity - borrowed_liquidity_wads.try_round_u64()?;
+            let collateral_mint_supply = Decimal::from(total_liquidity).try_mul(Rate::from_scaled_val(collateral_multiplier))?.try_round_u64()?;
+            let state = ReserveState {
+                borrowed_liquidity_wads,
+                available_liquidity,
+                collateral_mint_supply,
+                ..ReserveState::default()
+            };
+
+            state.collateral_exchange_rate()?;
+        }
+
+        #[test]
+        fn compound_interest(
+            slots_elapsed in 0..=SLOTS_PER_YEAR,
```

Comment:
I've only been testing up to one year. I can't imagine an active market would be untouched for that long so it felt like a good upper bound. It could be tested higher but you would run into compute unit limits first (I also use one year for integration tests)

Reply:
You're definitely right that the markets will be touched more often than that, but I was worried that something might blow out after many years.  Since this test reserve state always starts with a minimum `cumulative_borrow_rate_wads`, it would be useful to have the compound interest function tested for bigger numbers of `cumulative_borrow_rate_wads` as well, since it will always be growing exponentially.  My question might actually be more specific: do you know how many years of compounding you get until `cumulative_borrow_rate_wads` overflows?

Reply:
Ahh, sorry misunderstood. Good call, I split this test into two tests. One for compounding up to 1000 years ignoring potential borrowed value overflows, and one that compounds up to 1 year which tests borrowed value overflows. The reason for the split is that we can't really do anything if a rounded u64 borrowed value exceeds u64::MAX because that's the highest token value supported by the token program. If that increases, we'll need a new version of the lending program anyways

Reply:
Great, looks good!